### ISSUE TITLE
Infinite compilation on JDK11

### ISSUE THREAD

**Issue by ben-manes** [google/error-prone#1205]
When upgrading to JDK 11.01, the compilation never finishes but has high CPU activity. When removing ErrorProne, the compilation completes without a problem. I am unsure how to debug, e.g. to see what the javac process is doing, to provide more details.

Environment:
- JDK 11.01 (upgrading from 10+46)
- Mac OS X Mojave
- Gradle 5.0
- ErrorProne:
    - core: 2.3.2
    - javac: 9+181-r4173-1
    - plugin: 0.6
- Annotation Processors:
    - autoFactory: 1.0-beta6
    - autoValue: 1.6.3
    - autoValueBuilder: 2.9.1

Of course if I disable the annotation processors then it fails due to the missing generated classes, but does terminate. Removing `autoFactory` leads to the hang, while `autoValue` is used too frequently to get a hang prior. So there is a chance that it is related to annotation processing (such as due to Gradle's incremental support), but could be something else.

**Comment by Stephan202** [comment]
@ben-manes does running `jstack <pid>` give any clue?

**Comment by ben-manes** [comment]
Unfortunately not much that I can debug from, but maybe others can.

Running with `-d --no-daemon` I get 
```
14:12:03.701 [INFO] [org.gradle.api.internal.tasks.compile.JdkJavaCompiler] Compiling with JDK Java compiler API.
```

and the sections that look interesting are,
```
"C2 CompilerThread0" #5 daemon prio=9 os_prio=31 cpu=29238.27ms elapsed=211.65s tid=0x00007f87b683a800 nid=0x3f03 runnable  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE
   Compiling: 18724       4       com.squareup.javapoet.CodeWriter::emit (781 bytes)
```

```
Execution worker for ':'" #37 prio=5 os_prio=31 cpu=169492.65ms elapsed=204.29s tid=0x00007f87b630e800 nid=0x14d03 runnable  [0x000070000b233000]
   java.lang.Thread.State: RUNNABLE
	at java.nio.file.Files.provider(java.base@11.0.1/Files.java:100)
	at java.nio.file.Files.isDirectory(java.base@11.0.1/Files.java:2228)
	at com.sun.tools.javac.file.PathFileObject.<init>(jdk.compiler@11.0.1/PathFileObject.java:352)
	at com.sun.tools.javac.file.PathFileObject$DirectoryFileObject.<init>(jdk.compiler@11.0.1/PathFileObject.java:111)
	at com.sun.tools.javac.file.PathFileObject.forDirectoryPath(jdk.compiler@11.0.1/PathFileObject.java:102)
	at com.sun.tools.javac.file.JavacFileManager.getFileForOutput(jdk.compiler@11.0.1/JavacFileManager.java:872)
	at com.sun.tools.javac.file.JavacFileManager.getJavaFileForOutput(jdk.compiler@11.0.1/JavacFileManager.java:816)
	at jdk.internal.reflect.GeneratedMethodAccessor101.invoke(Unknown Source)
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@11.0.1/DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(java.base@11.0.1/Method.java:566)
	at org.gradle.api.internal.tasks.compile.reflect.SourcepathIgnoringInvocationHandler.invoke(SourcepathIgnoringInvocationHandler.java:64)
	at com.sun.proxy.$Proxy131.getJavaFileForOutput(Unknown Source)
	at com.sun.tools.javac.api.ClientCodeWrapper$WrappedJavaFileManager.getJavaFileForOutput(jdk.compiler@11.0.1/ClientCodeWrapper.java:298)
	at com.sun.tools.javac.processing.JavacFiler.createSourceOrClassFile(jdk.compiler@11.0.1/JavacFiler.java:500)
	at com.sun.tools.javac.processing.JavacFiler.createSourceFile(jdk.compiler@11.0.1/JavacFiler.java:426)
	at org.gradle.api.internal.tasks.compile.processing.IncrementalFiler.createSourceFile(IncrementalFiler.java:49)
	at net.autobuilder.core.AutoBuilderProcessor.write(AutoBuilderProcessor.java:176)
	at net.autobuilder.core.AutoBuilderProcessor.processTask(AutoBuilderProcessor.java:159)
	at net.autobuilder.core.AutoBuilderProcessor.doProcess(AutoBuilderProcessor.java:131)
	at net.autobuilder.core.AutoBuilderProcessor.process(AutoBuilderProcessor.java:109)
	at org.gradle.api.internal.tasks.compile.processing.DelegatingProcessor.process(DelegatingProcessor.java:62)
	at org.gradle.api.internal.tasks.compile.processing.AggregatingProcessor.process(AggregatingProcessor.java:50)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(jdk.compiler@11.0.1/JavacProcessingEnvironment.java:972)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(jdk.compiler@11.0.1/JavacProcessingEnvironment.java:888)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(jdk.compiler@11.0.1/JavacProcessingEnvironment.java:1214)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(jdk.compiler@11.0.1/JavacProcessingEnvironment.java:1326)
	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(jdk.compiler@11.0.1/JavaCompiler.java:1258)
	at com.sun.tools.javac.main.JavaCompiler.compile(jdk.compiler@11.0.1/JavaCompiler.java:936)
	at com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(jdk.compiler@11.0.1/JavacTaskImpl.java:104)
	at com.sun.tools.javac.api.JavacTaskImpl$$Lambda$200/0x0000000800c50c40.call(jdk.compiler@11.0.1/Unknown Source)
	at com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(jdk.compiler@11.0.1/JavacTaskImpl.java:147)
	at com.sun.tools.javac.api.JavacTaskImpl.doCall(jdk.compiler@11.0.1/JavacTaskImpl.java:100)
	at com.sun.tools.javac.api.JavacTaskImpl.call(jdk.compiler@11.0.1/JavacTaskImpl.java:94)
	at org.gradle.api.internal.tasks.compile.AnnotationProcessingCompileTask.call(AnnotationProcessingCompileTask.java:89)
	at org.gradle.api.internal.tasks.compile.ResourceCleaningCompilationTask.call(ResourceCleaningCompilationTask.java:57)
	at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:50)
```

I know that the `javac` is doing work as sampling the compilation thread shows updates to the stack, e.g.
```
"Execution worker for ':'" #37 prio=5 os_prio=31 cpu=407663.43ms elapsed=448.13s tid=0x00007f87b630e800 nid=0x14d03 runnable  [0x000070000b234000]
   java.lang.Thread.State: RUNNABLE
	at java.lang.Iterable.forEach(java.base@11.0.1/Iterable.java:75)
	at com.sun.tools.javac.code.DeferredCompletionFailureHandler$1.install(jdk.compiler@11.0.1/DeferredCompletionFailureHandler.java:68)
	at com.sun.tools.javac.code.DeferredCompletionFailureHandler.setHandler(jdk.compiler@11.0.1/DeferredCompletionFailureHandler.java:109)
	at com.sun.tools.javac.api.MultiTaskListener.finished(jdk.compiler@11.0.1/MultiTaskListener.java:127)
	at com.sun.tools.javac.main.JavaCompiler.enterTrees(jdk.compiler@11.0.1/JavaCompiler.java:1073)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.enterTrees(jdk.compiler@11.0.1/JavacProcessingEnvironment.java:1194)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.<init>(jdk.compiler@11.0.1/JavacProcessingEnvironment.java:1080)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.next(jdk.compiler@11.0.1/JavacProcessingEnvironment.java:1100)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(jdk.compiler@11.0.1/JavacProcessingEnvironment.java:1337)
	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(jdk.compiler@11.0.1/JavaCompiler.java:1258)
```

**Comment by cushon** [comment]
```
at com.sun.tools.javac.code.DeferredCompletionFailureHandler$1.uninstall(jdk.compiler@11.0.1/DeferredCompletionFailureHandler.java:82)
```

Maybe this is [JDK-8209055](https://bugs.openjdk.java.net/browse/JDK-8209055)?

Are you using `-XDcompilePolicy=simple`? If so, does removing that flag make a difference?

**Comment by ben-manes** [comment]
hmm, good idea on disabling all checks. I'll try that now.

I am not using anything fancy in this project, except for minor workarounds for module/jigsaw compatibility issues.

```gradle
tasks.withType(JavaCompile) {
  options.encoding = 'UTF-8'
  options.compilerArgs += [
    '-Xlint:all',
    '-Xlint:-processing',
    '--add-modules', 'java.compiler',
  ]
  options.errorprone {
    disableWarningsInGeneratedCode = true
    excludedPaths = "${buildDir}/generated-sources/.*"
  }

  afterEvaluate {
    configurations.patchJsr305.each {
      options.compilerArgs += [ '--patch-module', "java.xml.ws.annotation=${it}" ]
    }
  }
}
```

**Comment by ben-manes** [comment]
No luck with disabling all checks. I also tried updating to `2.3.3-SNAPSHOT` which also hangs.

**Comment by Stephan202** [comment]
Based on the posted config it seems you're not compiling with `-XDcompilePolicy=simple`. Maybe add that? (I.e. the inverse of @cushon's suggestion; admittedly I don't know whether this suggestion makes sense, but the documentation _does_ recommend using this flag.)

**Comment by cushon** [comment]
I'm still suspicious of JDK-8209055.

I think `-XDcompilePolicy=simple` is baked in to the gradle plugin:
https://github.com/tbroyer/gradle-errorprone-plugin/blob/21ac96affce535420cf8266e8095a16047eb3e82/src/main/kotlin/net/ltgt/gradle/errorprone/ErrorPronePlugin.kt#L172

Error Prone is supposed to throw an exception if that flag isn't set: https://github.com/google/error-prone/blob/20bf092117ec70e14dfdd125d49f36c5a3f5bc0c/check_api/src/main/java/com/google/errorprone/BaseErrorProneJavaCompiler.java#L164

**Comment by ben-manes** [comment]
I did try adding `-XDcompilePolicy=simple` which hung, though I see now you're right it's probably set implicitly. I tried adding `doFirst { println options.compilerArgs }` and didn't see it, and would have to dig in further to see if it can be manipulated / exposed somehow

```
> Task :compileJava
[-Xlint:all, -Xlint:-processing, --add-modules, java.compiler, --patch-module, java.xml.ws.annotation=/Users/ben/.gradle/caches/modules-2/files-2.1/com.google.code.findbugs/jsr305/3.0.2/25ea2e8b0c338a877313bd4672d3fe056ea78f0d/jsr305-3.0.2.jar]
```

**Comment by ben-manes** [comment]
I tried disabling error prone and adding `-XDcompilePolicy=simple` to `options.compilerArgs` and it did not hang, if that helps any.

**Comment by cushon** [comment]
> I tried disabling error prone and adding `-XDcompilePolicy=simple` to options.compilerArgs and it did not hang, if that helps any.

That weakens my theory but I don't think it's conclusive, my understanding of the bug is that it might be exacerbated by running Error Prone.

Can you try running Error Prone and modifying the gradle plugin to use `-XDcompilePolicy=byfile` instead of `simple`?

Alternately, you could try a JDK 12 early access build, since it includes the fix for that bug.

**Comment by ben-manes** [comment]
Okay, I'll try JDK 12 since that's faster than making a custom build of the plugin.

P.S. Thanks for the help so far @Stephan202 & @cushon

**Comment by ben-manes** [comment]
Java 12 worked! 😅

**Comment by tbroyer** [comment]
> @tbroyer is there a good way to customize `-XDcompilePolicy=simple` when using Error Prone with gradle?

Not what I'd call a "good way", but it should be possible (using the Kotlin DSL, as my Groovy is rather bad):
```kotlin
tasks.withType<JavaCompile>().configureEach {
  // prevent net.ltgt.errorprone plugin from adding the compiler args
  options.errorprone.isEnabled = false
  // add the `-Xplugin:ErrorProne` argument with Error Prone options, but without the -XDcompilePolicy
  options.compilerArgumentProviders.add(CommandLineArgumentProvider {
    listOf("-Xplugin:ErrorProne ${options.errorprone}")
  })
}
```

(fwiw, @ben-manes, you might want to use `compilerArgumentProviders` instead of your `afterEvaluate` for your `--patch-module`; declaring the configuration as a `@Nested` property in the provider's class)

@cushon should I add a configurable property to the plugin to allow setting the `-XDcompilePolicy`? (or at a minimum toggling whether the plugin adds it, so one could add its own) or it's not necessary as this is fixed already? (and should be backported to javac 11 at some point)

**Comment by cushon** [comment]
@jbrinegar can you provide a repro?

**Comment by jbrinegar** [comment]
@cushon unfortunately, I cannot provide a reproducing test case.  I did try a handful of open source projects on 11 + errorprone, but it does look like this is related to annotation processors.  For now, I am working around this by just not allowing errorprone to run on the impacted module.

I did confirm that JDK-8209055, when applied to jdk11u, does fix the problem entirely.  It should be nominated for backporting.
